src/orderbook.ts

```diff
@@ -99,18 +139,93 @@ export class Orderbook {
       this.addAsk(ask);
     });
   }
+
+  /**
+   * Computes the transitive closure of this orderbook (e.g. ETH/DAI) with another one (e.g. DAI/USDC).
+   * Throws if the orderbooks cannot be combined (baseToken is not equal to quoteToken)
+   * @param orderbook The orderbook for which the transitive closure will be computed
+   * @returns A new instance of an orderbook representing the resulting closure.
+   */
+  transitiveClosure(orderbook: Orderbook) {
+    if (orderbook.baseToken != this.quoteToken) {
+      throw new Error(
+        `Cannot compute transitive closure of ${this.pair()} orderbook and ${orderbook.pair()} orderbook`
+      );
+    }
+
+    const ask_closure = this.transitiveAskClosure(orderbook);
+
+    // Since bids are the asks of the inverted orderbook, computing transitive closure of bids is equivalent to
+    // 1) inverting both orderbooks
+    // 2) computing the transitive ask closure on the inverses
+    // 3) re-inverting the result
+    const bid_closure = orderbook
+      .inverted()
+      .transitiveAskClosure(this.inverted())
+      .inverted();
+
+    ask_closure.add(bid_closure);
+    return ask_closure;
+  }
+
+  private transitiveAskClosure(orderbook: Orderbook) {
+    const result = new Orderbook(this.baseToken, orderbook.quoteToken);
+
+    // Create a copy here so original orders stay untouched
+    const left_asks = Array.from(this.asks.values());
```

Comment:
This might not be a deep clone, in that modifying the fractions might modify it in both arrays. I'm not sure though - my JS knowledge is not super great.

Reply:
You are right that this is not a deep clone.

All the methods on fraction return new instances and by making numerator/denominator private they can not be modified from outside.

However, you are right that this is not enforced by the compiler for the future. The way to do this would be to implement a clone method on all our datatypes and invoke this for each Offer. Not sure this is worth it.

Reply:
You are right, I don't think its worth it. You should mark the Fraction properties as `readonly`. That way they can only be assigned in the constructor and the compiler guarantees they are not mutated.

Reply:
`reduce` reassigns the numerator and denominator by dividing them with the gcd, thus readonly doesn't work. Moreover, since BN also has some in-place methods (e.g. imul, iadd, but ironically not idiv at least in the typescript bindings) the values can be mutated even if they are not reassigned.

Rust is a lot stronger when it comes to type/mutation safety.

Reply:
It has `idivn` though :stuck_out_tongue: 

And really great points, Its nice the way it is!